src/main/kotlin/org/rust/ide/ssr/RsMatchingVisitor.kt

```diff
@@ -0,0 +1,258 @@
+                match(o.vis, field.vis) &&
+                matchIdentifier(o.identifier, field.identifier) &&
+                match(o.typeReference, field.typeReference)
+    }
+
+    override fun visitTupleFields(o: RsTupleFields) {
+        val fields = getElement<RsTupleFields>() ?: return
+        myMatchingVisitor.result = myMatchingVisitor.matchSequentially(o.tupleFieldDeclList, fields.tupleFieldDeclList)
+    }
+
+    override fun visitTupleFieldDecl(o: RsTupleFieldDecl) {
+        val field = getElement<RsTupleFieldDecl>() ?: return
+        myMatchingVisitor.result =
+            matchOuterAttrList(o.outerAttrList, field.outerAttrList) &&
+                match(o.vis, field.vis) &&
+                match(o.typeReference, field.typeReference)
+    }
+
+    override fun visitTypeReference(o: RsTypeReference) {
+        val typeReference = getElement<RsTypeReference>() ?: return
+        // TODO: implement individual type references
+        myMatchingVisitor.result = matchTextOrVariable(o, typeReference)
+    }
+
+    override fun visitRefLikeType(o: RsRefLikeType) {
+        val refType = getElement<RsRefLikeType>() ?: return
+        myMatchingVisitor.result = (o.mut == null) == (refType.mut == null) &&
+                (o.const == null) == (refType.const == null) &&
+                (o.mul == null) == (refType.mul == null) &&
+                match(o.typeReference, refType.typeReference) &&
+                match(o.lifetime, refType.lifetime)
+    }
+
+    override fun visitLifetime(o: RsLifetime) {
+        val lifetime = getElement<RsLifetime>() ?: return
+        myMatchingVisitor.result = matchTextOrVariable(o, lifetime)
+    }
+
+    override fun visitVis(o: RsVis) {
+        val vis = getElement<RsVis>() ?: return
+        myMatchingVisitor.result = matchTextOrVariable(o, vis)
+    }
+
+    override fun visitInnerAttr(o: RsInnerAttr) {
+        val attr = getElement<RsInnerAttr>() ?: return
+        myMatchingVisitor.result = match(o.metaItem, attr.metaItem)
+    }
+
+    override fun visitOuterAttr(o: RsOuterAttr) {
+        val attr = getElement<RsOuterAttr>() ?: return
+        myMatchingVisitor.result = match(o.metaItem, attr.metaItem)
+    }
+
+    override fun visitMetaItem(o: RsMetaItem) {
+        val metaItem = getElement<RsMetaItem>() ?: return
+        myMatchingVisitor.result = match(o.compactTT, metaItem.compactTT) &&
+            match(o.litExpr, metaItem.litExpr) &&
+            match(o.metaItemArgs, metaItem.metaItemArgs) &&
+            match(o.path, metaItem.path)
+    }
+
+    override fun visitMetaItemArgs(o: RsMetaItemArgs) {
+        val metaItemArgs = getElement<RsMetaItemArgs>() ?: return
+        myMatchingVisitor.result = myMatchingVisitor.matchInAnyOrder(o.metaItemList, metaItemArgs.metaItemList)
+    }
+
+    override fun visitPathType(o: RsPathType) {
+        val path = getElement<RsPathType>() ?: return
+
+        // Match the last path component
```

Comment:
What is motivation for matching paths this way? Personally I think path pattern should match path if they both resolve to the same item. Note that it can be done without `ImportCandidatesCollector` like this:
```kotlin
private fun resolvesToSameItem(templatePath: RsPath, treePath: RsPath): Boolean {
    val path = RsCodeFragmentFactory(treePath.project)
        .createPath(templatePath.text, treePath, treePath.pathParsingMode, treePath.allowedNamespaces()) ?: return false
    return path.reference?.resolve() == treePath.reference?.resolve()
}
```

---

Here are a few examples to think of and add tests:

1. Path pattern is fully-qualified

Pattern:
```rust
const $name$: crate::mod1::Foo = 0;
```
Code:
```rust
const x1: mod1::Foo = 0;
const x2: crate::mod1::Foo = 0;
mod inner {
    use crate::mod1::Foo;
    const x3: Foo = 0;
}
mod mod1 {
    pub struct Foo;
}
```

2. Type aliases

Pattern:
```rust
const $name$: i32 = 0;
```
Code:
```rust
const x: T = 0;
type T = i32;
```

Reply:
> I think path pattern should match path if they both resolve to the same item

Do we really want it? As I understand the concept of SSR, it's a way how to write patterns similar to regular expressions but for wider scope of languages (not only regular languages). I.e. it should [mostly] use syntax rules. @yopox am I right? If so, I'd like to keep Rust declarative macros model in mind during matching

But name resolution is a semantic rule.
Two more points not to use name resolution:
- it may cause performance issues because of more work
- it doesn't allow creating syntax only rules. Of course, otherwise, it won't be possible to create semantic rules but not sure that we really need them since they may be too complex for users

Reply:
Maybe semantic rules should be expressed as modifiers

Reply:
I would definitely want it to be able to match different expressions which resolve to the same one. There are too many ways to write equivalent expressions in Rust. A method could be called as `foo.bar()`, `(&foo).bar()`, `Foo::bar(&foo)`, `Trait::bar(&foo)`, `path::Trait::bar(&foo)`, `a::b::ReExport::bar(&foo)` etc. If SSR is able to do only literal matches, it would be useless.  Particularly for SSR-based lints.

On the other hand, sometimes the literal tokens are exactly what I want to match, so there should be an option to do either.

Reply:
@dima74 We can not resolve the path from the template to an item :(

Consider this template: `struct $S$(bevy::$x$::HashMap)`, or simply `struct $S1$(bevy::utils::$S2$)`, to get these matching we have to go through the matching visitor machinery.

A working approach I have tried is:

1. Matching the last part of the path
2. Check if the rest is matching
3. If 1 is OK but not 2:
    - Collect import paths for the code path
    - Build `RsPath` with the fully qualified import path
    - Check if any is matching

We must use resolve because it is absurd not to match `crate::foo::Foo` with `Foo` in the code (where it is the same reference). It will look if we can have a language specific option in the SSR dialog to disable the use of resolve. If not we will be able to do `crate::foo::$X$` with "Don't resolve paths" modifier on `$X$`.